fixes compatibility with macos big sur.

TheSecEng opened the pull request:
MacOS Big Sur had dynamic linker changes. Shared libraries are now stored only in `/System/Library/dyld/dyld_shared_cache_x86_64` and no longer present on the actual file system. This changes makes the `find_library` not return the paths of the libraries such as `find_library("ssl")`

wbond commented on oscrypto/_openssl/_libssl_ctypes.py:
```diff
@@ -20,6 +20,9 @@
 if libssl_path is None:
     libssl_path = find_library('ssl')
+    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
```
Same as libcrypto

wbond commented on oscrypto/_openssl/_libssl_cffi.py:
```diff
@@ -24,6 +24,9 @@
 if libssl_path is None:
     libssl_path = find_library('ssl')
+    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
```
Same as libcrypto

wbond commented on oscrypto/_openssl/_libcrypto_cffi.py:
```diff
@@ -27,10 +27,12 @@
 if libcrypto_path is None:
     libcrypto_path = find_library('crypto')
+    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
```
So here we should check if `libcrypto_path` is empty. The code below had a bug because always assumed a system-level location, but some users have run it in conda, etc.

TheSecEng replied:
You mean like this? Still keeping the logic for MacOS 10.15+? 

```diff
libcrypto_path = _backend_config().get('libcrypto_path')
if libcrypto_path is None:
    libcrypto_path = find_library('crypto')
+    if sys.platform == 'darwin' and not libcrypto_path:
+        libcrypto_path == '/usr/lib/libcrypto.42.dylib'
-    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
-    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
-        libcrypto_path == '/usr/lib/libcrypto.42.dylib'
-    # if we are on macOS 10.15+, we want to strongly version libcrypto since unversioned libcrypto has a non-stable ABI
-    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 15) and \
-            libcrypto_path == '/usr/lib/libcrypto.dylib':
-        libcrypto_path = libcrypto_path.replace('libcrypto.dylib', 'libcrypto.42.dylib')
if not libcrypto_path:
    raise LibraryNotFoundError('The library libcrypto could not be found')
```

wbond replied:
No, just:

```suggestion
    if not libcrypto_path and sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
```

That way we only hard code it for it being missing on 10.16. We know that version will be there on 10.16, but it doesn't exist on older versions of macOS.

wbond commented on oscrypto/_openssl/_libcrypto_ctypes.py:
```diff
@@ -26,8 +26,11 @@
 if libcrypto_path is None:
     libcrypto_path = find_library('crypto')
+    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
```
Same as cffi backend.

FichteFoll commented on oscrypto/_openssl/_libcrypto_cffi.py:
```diff
@@ -27,10 +27,13 @@
+        libcrypto_path == '/usr/lib/libcrypto.42.dylib'
     # if we are on macOS 10.15+, we want to strongly version libcrypto since unversioned libcrypto has a non-stable ABI
-    if sys.platform == 'darwin' and list(map(int, platform.mac_ver()[0].split('.')))[1] >= 15 and \
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 15) and \
```
Should probably become an `elif`. While it running for 10.16+ doesn't change because the last condition won't apply, this seems more fragile than needs be.

FichteFoll commented on oscrypto/_openssl/_libcrypto_ctypes.py:
```diff
@@ -26,8 +26,12 @@
+        libcrypto_path == '/usr/lib/libcrypto.42.dylib'
     # if we are on macOS 10.15+, we want to strongly version libcrypto since unversioned libcrypto has a non-stable ABI
-    if sys.platform == 'darwin' and list(map(int, platform.mac_ver()[0].split('.')))[1] >= 15 and \
+    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 15) and \
```
Same as above.

FichteFoll commented on oscrypto/_openssl/_libssl_cffi.py:
```diff
@@ -24,6 +24,10 @@
+            tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
+        libssl_path = '/usr/lib/libssl.44.dylib'
     # if we are on macOS 10.15+, we want to strongly version libssl since unversioned libcrypto has a non-stable ABI
     if sys.platform == 'darwin' and list(map(int, platform.mac_ver()[0].split('.')))[1] >= 15 and \
```
Same as above.

FichteFoll commented on oscrypto/_openssl/_libssl_ctypes.py:
```diff
@@ -20,6 +20,10 @@
+            tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
+        libssl_path = '/usr/lib/libssl.44.dylib'
     # if we are on macOS 10.15+, we want to strongly version libssl since unversioned libcrypto has a non-stable ABI
     if sys.platform == 'darwin' and list(map(int, platform.mac_ver()[0].split('.')))[1] >= 15 and \
```
Same as above.

FichteFoll commented:
Actually, seeing how frequently this pattern is used, it seems reasonable to factor it out into a function of some sort.

TheSecEng commented:
> Actually, seeing how frequently this pattern is used, it seems reasonable to factor it out into a function of some sort.

Have something in mind? This? Welcomed suggestions, I am on the road currently.

```python
def get_library(library, fallback, unverisoned):
    ...
```

TheSecEng commented:
Not sure I like this

```python
    library = find_library(name)
    fallback_path = '/usr/lib/{}'.format(fallback)
    # If we are on macOS 10.16+, find_library doesn't work, so we set a static path
    if not library and sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
        library == fallback
    # if we are on macOS 10.15+, we want to strongly version libcrypto since unversioned libcrypto has a non-stable ABI
    if sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 15) and \
            library == unversioned_path:
        # libcrypto.42.dylib is in libressl-2.6 which as a OpenSSL 1.0.1-compatible API
        library = fallback
```

TheSecEng commented:
@FichteFoll mind helping out with this test issue? I I am guessing I need to fix something in the tests themselves, unless my logic is wrong.


```
Running tests (via coverage.py)
Traceback (most recent call last):
  File "run.py", line 8, in <module>
    run_task()
  File "/Users/terminal/.projects/oscrypto/dev/_task.py", line 162, in run_task
    result = run(*args, **kwargs)
  File "/Users/terminal/.projects/oscrypto/dev/ci.py", line 49, in run
    tests_result = run_coverage(ci=True)
  File "/Users/terminal/.projects/oscrypto/dev/coverage.py", line 57, in run
    result = run_tests(ci=ci)
  File "/Users/terminal/.projects/oscrypto/dev/tests.py", line 61, in run
    for test_class in test_classes():
  File "/Users/terminal/.projects/oscrypto/tests/__init__.py", line 136, in test_classes
    from .test_kdf import KDFTests
  File "/Users/terminal/.projects/oscrypto/tests/test_kdf.py", line 7, in <module>
    from oscrypto import kdf, _pkcs5
  File "/Users/terminal/.projects/oscrypto/oscrypto/kdf.py", line 9, in <module>
    from .util import rand_bytes
  File "/Users/terminal/.projects/oscrypto/oscrypto/util.py", line 13, in <module>
    from ._mac.util import rand_bytes
  File "/Users/terminal/.projects/oscrypto/oscrypto/_mac/util.py", line 208, in <module>
    from .._openssl._libcrypto import libcrypto
  File "/Users/terminal/.projects/oscrypto/oscrypto/_openssl/_libcrypto.py", line 9, in <module>
    from ._libcrypto_cffi import (
  File "/Users/terminal/.projects/oscrypto/oscrypto/_openssl/_libcrypto_cffi.py", line 7, in <module>
    from ..util import get_library
ImportError: cannot import name 'get_library' from 'oscrypto.util' (/Users/terminal/.projects/oscrypto/oscrypto/util.py)
```

wbond commented:
I believe you have an include loop since the root `util.py` is importing the platform-specific one to get `rand_bytes()`.

This helper function really shouldn’t be part of the public API, so it should go in a file with an `_`, probably `_ffi.py`.

FichteFoll commented on oscrypto/_openssl/_libcrypto_cffi.py:
```diff
@@ -4,8 +4,8 @@
 from .. import _backend_config
-from ..util import get_library
 from .._errors import pretty_message
+from .._ffi import get_library
 from .._ffi import register_ffi
```
```suggestion
from .._ffi import get_library, register_ffi
```

FichteFoll commented on oscrypto/_openssl/_libcrypto_ctypes.py:
```diff
@@ -6,8 +6,8 @@
 from .. import _backend_config
-from ..util import get_library
 from .._errors import pretty_message
+from .._ffi import get_library
 from .._ffi import FFIEngineError
```
```suggestion
from .._ffi import FFIEngineError, get_library
```

FichteFoll commented on oscrypto/_openssl/_libssl_cffi.py:
```diff
@@ -2,7 +2,7 @@
 
 from .. import _backend_config
-from ..util import get_library
+from .._ffi import get_library
 from .._ffi import register_ffi
```
```suggestion
from .._ffi import get_library, register_ffi
```

FichteFoll commented on oscrypto/_openssl/_libssl_ctypes.py:
```diff
@@ -4,7 +4,7 @@
 
 from .. import _backend_config
-from ..util import get_library
+from .._ffi import get_library
 from .._ffi import FFIEngineError
```
```suggestion
from .._ffi import FFIEngineError, get_library
```

FichteFoll commented on oscrypto/_ffi.py:
```diff
@@ -386,6 +390,31 @@ def callback(library, signature_type, func):
     engine = 'ctypes'
 
 
+def get_library(name, unversioned, fallback):
+    """
+    Retrieves the C library, falling back to a specified library if one is not found
+
+    :param name:
+        The library to search the system for
+
+    :param unversioned:
+        The unversioned library we don't want to use
+
+    :param fallback:
+        Fallback library when we don't find a suitable library to use
+
+    :return:
+        Path to the library
+    """
+    library = find_library(name)
+    if not library and sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 16):
+        library == fallback
+    elif sys.platform == 'darwin' and tuple(map(int,  platform.mac_ver()[0].split('.'))) >= (10, 15) and \
+            library == unversioned:
+        library = fallback
+    return library
```
I suggest this instead:
```suggestion
def get_library(name, dylib_name, version):
    """
    Retrieve the C library path with special handling for macOS.
    
    :param name:
        The library to search the system for.
    :param unversioned:
        The expected unversioned dylib name.
    :param version:
        dylib version override.
        Preferred for macOS 15.15+
        and used as a fallback for 10.16 where `find_library` doesn't work.
    :return:
        Path to the library.
    """
    library = find_library(name)
    if sys.platform == 'darwin':
        unversioned = '/usr/lib/{}'.format(dylib_name)
        versioned = unversioned.replace('.dylib', '.{}.dylib'.format(version))
        mac_ver = tuple(map(int, platform.mac_ver()[0].split('.')))
        if not library and mac_ver >= (10, 16):
            # On macOS 10.16+, find_library doesn't work, so we set a static path
            library = versioned
        elif mac_ver >= (10, 15) and library == unversioned:
            # On macOS 10.15+, we want to strongly version since unversioned libcrypto has a non-stable ABI
            library = versioned
    return library
```
1. Fixed bug where the first condition's body was an equation, not an assignment.
1. Changed function signature for less repitition and adjusted parameter names to more strongly reflect their purpose (`fallback` was a concious choice).
1. Restructured conditions and extracted mac version into a variable.
1. Preserved relevant comments in docstring. The comments explaining the particular version choices (for OpenSSL 1.0.1 compliance) are still lost, howeer.

michaelblyons commented:
`/backlink`-ing to https://github.com/wbond/package_control/pull/1477

wbond commented on oscrypto/_ffi.py:
```diff
@@ -386,6 +390,35 @@ def callback(library, signature_type, func):
+    """
+    library = find_library(name)
+    if sys.platform == 'darwin':
+        unversioned = '/usr/lib/{}'.format(dylib_name)
```
From the tests it would appear this code doesn't work on Python 2.6. The test of the code base pretty much uses `%`, so I'd probably stick with that.

Same for the line below.

FichteFoll replied:
Indeed, empty formatting specifiers without an index are only supported on Python 3.

michaelblyons replied:
> Indeed, empty formatting specifiers without an index are only supported on Python 3.

That can't be true, can it? I was definitely using those on 2.7.x with IronPython, and maybe 2.6.x.

michaelblyons replied:
Okay, so not 2.6, but [the docs](https://docs.python.org/2.7/library/string.html#formatstrings) _do_ say 2.7:

> _Changed in version 2.7:_ The positional argument specifiers can be omitted for `str.format()` and `unicode.format()`, so `'{} {}'` is equivalent to `'{0} {1}'`, `u'{} {}'` is equivalent to `u'{0} {1}'`.

FichteFoll replied:
Ah yes, I made that mistake a while back already. Pretty much all the Python 2 code I ever wrote was for Python 2.6, so I mix some of these limitations.

wbond commented:
The error seen with Python 3.7 on Mac can be resolved by rebasing on master, although you may not see it every time the tests are run.

wbond commented:
Apparently installing PyPy via homebrew is broken, so I'm just going to ignore that error for now.

Thanks for the work on this!
